File: cpp/208.cpp
#include <string>
using namespace std;

class TrieNode {
public:
  bool exists;
  TrieNode **children;
  // Initialize your data structure here.
  TrieNode() {
    children = new TrieNode*[26](); // () for init
    exists = false;
  }
};

class Trie {

private:
  TrieNode *root;

  TrieNode* search(string word, bool create){
    TrieNode *p = root;
    for (char c : word) {
      int id = c - 'a';
      if (p->children[id] == nullptr){
        if(create)
          p->children[id] = new TrieNode();
        else
          return nullptr;
      }
      p = p->children[id];
    }
    return p;
  }

public:
  Trie() { root = new TrieNode(); }

  // Inserts a word into the trie.
  void insert(string word) {
    search(word, true)->exists = true;
  }

  // Returns if the word is in the trie.
  bool search(string word) {
    TrieNode *p = search(word, false);
    return p && p->exists;
  }

  // Returns if there is any word in the trie
  // that starts with the given prefix.
  bool startsWith(string prefix) {
    return search(prefix, false) != nullptr;
  }
};

File: python/208.py
TrieNode = dict

class Trie(object):

    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        """
        Inserts a word into the trie.
        :type word: str
        :rtype: void
        """
        p = self.root
        for c in word:
            if c not in p:
                p[c] = {}
            p = p[c]
        p[True] = True

    def search(self, word):
        """
        Returns if the word is in the trie.
        :type word: str
        :rtype: bool
        """
        p = self.root
        for c in word:
            if c not in p:
                return False
            p = p[c]
        return True in p

    def startsWith(self, prefix):
        """
        Returns if there is any word in the trie
        that starts with the given prefix.
        :type prefix: str
        :rtype: bool
        """
        p = self.root
        for c in prefix:
            if c not in p:
                return False
            p = p[c]
        return True

# Your Trie object will be instantiated and called as such:
trie = Trie()
trie.insert("somestring")
print(trie.search("key"))
print(trie.search("somestring"))
print(trie.search("somestrin"))
print(trie.startsWith("somestring "))
